### ISSUE TITLE
YOLO v8 custom loss implementation on the classification task

### ISSUE THREAD

**Issue by soumen25-git** [ultralytics/ultralytics#22629]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Hello team,

I am using YOLO v8 for my object detection task on a custom dataset. I want to incorporate domain knowledge of my dataset in the classification loss function. Therefore, I am trying to modify the v8DetectionLoss class forward function. However I have some confusion in the below points.

1. The dimension of pred and target tensor are (16, 8400, 80). How does the target tensor arranged following this dimension? 
2. And what is the significance of 8400 dimension?

Any help or explanation will be deeply appreciated.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @soumen25-git, thank you for your interest in Ultralytics 🚀! This is an automated reply to help you get started while an Ultralytics engineer reviews your issue and will assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (as it appears here with custom loss modifications), please include the following to help us understand and reproduce your setup:
- Exact `ultralytics` version and environment details (OS, Python, PyTorch, CUDA)
- A minimal code snippet showing your `v8DetectionLoss.forward` modifications and how you pass `pred` and `target`
- A small sample of your dataset or a synthetic batch demonstrating shapes and contents of `pred` and `target`
- Relevant training logs and a brief description of the task (detection vs classification) and your intended custom loss behavior

Also verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) ✅

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Before proceeding, please upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Once you share the requested details, we’ll help clarify shapes and targets in your setup and guide you on safely integrating custom loss logic 💡

**Comment by glenn-jocher** [comment]
Thanks for the questions—here’s how to read those shapes for YOLO’s detection head: the tensors are `[batch, anchors, classes] = (16, 8400, 80)`, where 8400 is the total anchor points across the three feature maps at strides 8/16/32 for a 640×640 input: 80×80 + 40×40 + 20×20 = 8400; the targets are built per anchor by the task‑aligned assigner so each row is all zeros for background and one non‑zero entry for the matched class (masked by `fg_mask`), see the classification term in `v8DetectionLoss.__call__` and how `target_scores` are formed in `get_targets` for details. Refer to the docs for the loss pipeline in [v8DetectionLoss (loss.py)](https://docs.ultralytics.com/reference/utils/loss/) and target construction in [get_targets (tal.py)](https://docs.ultralytics.com/reference/utils/tal/).

If you’re customizing the classification loss, this is the line you’d replace:
```python
# inside v8DetectionLoss.__call__()
loss[1] = self.bce(pred_scores, target_scores.to(dtype)).sum() / target_scores_sum
```
Quick check for 8400 anchors:
```python
imgsz = 640
strides = [8, 16, 32]
print(sum((imgsz // s) * (imgsz // s) for s in strides))  # 8400
```

**Comment by soumen25-git** [comment]
Thank you @glenn-jocher for your explanation. It really helped me. Good day.

**Comment by glenn-jocher** [comment]
Glad it helped—if you’re all set please close the issue; if you want to encode domain knowledge, replace the BCE classification term in `v8DetectionLoss.__call__` with your own criterion (for example, see the [VarifocalLoss reference](https://docs.ultralytics.com/reference/utils/loss/#ultralytics.utils.loss.VarifocalLoss)).

**Comment by Chuong98CC** [comment]
oh, why we use BCE loss instead of Focal Loss or VariFocal Loss? I am reading the code and surprised since I always think Focal Loss as the de factor for Object Detection

**Comment by glenn-jocher** [comment]
Good question—YOLOv8 uses plain BCE for classification because, with the current label assignment and weighting, it gives a good balance of stability, speed, and accuracy without the extra hyperparameters and tuning overhead that Focal/Varifocal Loss introduce; that said, both `FocalLoss` and `VarifocalLoss` are implemented in `ultralytics.utils.loss`, so if your dataset is heavily class‑imbalanced you can swap that BCE line in `v8DetectionLoss.__call__` for one of those and compare, and the [Focal Loss glossary](https://www.ultralytics.com/glossary/focal-loss) has more background on when such losses help.

**Comment by Chuong98CC** [comment]
Thank you.

**Comment by Chuong98CC** [comment]
Thank you.

**Comment by glenn-jocher** [comment]
You’re welcome—if you later experiment with alternative classification losses and have results or run into issues, please open a new issue so the team can take a look.
